File: ___HappyCityScripts/EginPlugins/Connect/HttpConnect.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;
using System.Text.RegularExpressions;

public class HttpConnect {
	
	private static HttpConnect _instance;
	public static HttpConnect Instance {
		get {
			if(_instance == null) {
				_instance = new HttpConnect();
			}
			return _instance;
		}
	}

	public WWW HttpRequestAli (string url) {
		EginTools.Log(url);		

		WWW www = new WWW (url);
		return www;
	}
	
	public WWW HttpRequest (string url, WWWForm form) {	
		EginTools.Log(url);	

		if (form == null) { form = new WWWForm(); }

		if(form != null){
			long ms=EginTools.nowMinis() ;
			long mms = ms + EginTools.localBeiJingTime;
			string ccode = EginTools.encrypTime (mms.ToString());
 			form.AddField("client_code", ccode);
		}

		
		WWW www = (form == null) ? new WWW(url) : new WWW(url, form);
		return www;
	}
	public WWW HttpRequestWithSession (string url, WWWForm form) {	
		EginTools.Log(url);	
		
		string cookie = (EginUser.Instance.session != null)?EginUser.Instance.session:"";
    	Hashtable requestHeaders = new Hashtable();
    	requestHeaders.Add( "Cookie", cookie);


        Dictionary<string, string> headers = new Dictionary<string, string>();
        headers.Add("Cookie", cookie);
		
		if (form == null) { form = new WWWForm(); }
		form.AddField("Cookie", cookie);
	
		if(form != null){
			long ms=EginTools.nowMinis() ;
			long mms = ms + EginTools.localBeiJingTime;
			string ccode = EginTools.encrypTime (mms.ToString());
			
			form.AddField("client_code", ccode);
		}

        WWW www = new WWW(url, form.data, headers);
		return www;
	}

    public HttpResult BaseResult(WWW www)
    {
        return BaseResult(www,true);
    }

    public HttpResult BaseResult (WWW www, bool isSwitchUrl) {
		HttpResult result = new HttpResult();

		if(www.error != null) {
 			EginTools.Log("Http Failed: " + www.error); //UnityEngine.Debug.Log("CK : ------------------------------ error = " + www.error);
            if (isSwitchUrl)
            {
                PlatformGameDefine.playform.swithWebHostUrl();//切换网页IP
                result.isSwitchHost = true;
            }
		}else {
            string tempResultStr = www.text.Trim(); //
			UnityEngine.Debug.Log("CK : ------------------------------ text = " + tempResultStr);

            EginTools.Log("base:::"+tempResultStr);
			
			JSONObject resultObj = new JSONObject(tempResultStr);
			if (resultObj.type == JSONObject.Type.NULL) {

                if (isSwitchUrl)
                {
                    PlatformGameDefine.playform.swithWebHostUrl();//切换网页IP
                    result.isSwitchHost = true;
                }
				result.resultObject = ZPLocalization.Instance.Get("HttpConnectError");
			}else {
				if(resultObj){
					string resultType =null ;//= resultObj["result"].str;
					if(resultObj["result"] != null){
						resultType = resultObj["result"].str;
					}
					if ("ok".Equals(resultType)) {
						result.resultObject = resultObj["body"];
						result.resultType = HttpResult.ResultType.Sucess;
					}else {
						if(resultObj["body"] !=null){
							result.resultObject = Regex.Unescape(resultObj["body"].str);
						}else{
							result.resultObject = null;
						}

					}
				}else{
					Debug.Log("解析出错了---HttpConnect.cs -- 60");
				}

			}
		}
		return result;
	}
	
	/* ------ Register ------ */
	public HttpResult RegisterResult (WWW www) {
		HttpResult result = new HttpResult();
		if(www.error != null) {
			EginTools.Log("Http Failed: " + www.error);
		}else {
			string tempResultStr = www.text.Trim();
			EginTools.Log(tempResultStr);
			
			if ("ok".Equals(tempResultStr)) {
				result.resultType = HttpResult.ResultType.Sucess;
			}else if(tempResultStr.Length > 0) {
				JSONObject resultObj = new JSONObject(tempResultStr);
				if (resultObj.type == JSONObject.Type.NULL) {
					result.resultObject = Regex.Unescape(tempResultStr);
				}else {
					result.resultObject = Regex.Unescape(resultObj["error"].str);
				}
			}else {
				result.resultObject = ZPLocalization.Instance.Get("HttpConnectError");
			}
		}
		return result;
	}
	
	/* ------ Login ------ */
	public HttpResult UserLoginResult (WWW www) {
		HttpResult result = BaseResult(www);
		if(HttpResult.ResultType.Sucess == result.resultType) {
			string session = "";
			if(www.responseHeaders.ContainsKey("SET-COOKIE")) {
				session = www.responseHeaders["SET-COOKIE"];
			}
			Dictionary<string, string> resultDict = ((JSONObject)result.resultObject).ToDictionary();
			
			EginUser.Instance.InitUserWithDict(resultDict, session);
			EginUser.Instance.isGuest = false;
		}
		return result;
	}
	
	public HttpResult GuestLoginResult (WWW www) {
		HttpResult result = BaseResult(www);
		if(HttpResult.ResultType.Sucess == result.resultType) {
			string session = "";
			if(www.responseHeaders.ContainsKey("SET-COOKIE")) {
				session = www.responseHeaders["SET-COOKIE"];
			}
			Dictionary<string, string> resultDict = ((JSONObject)result.resultObject).ToDictionary();
			
			EginUser.Instance.InitGuestWithDict(resultDict, session);
			EginUser.Instance.isGuest = true;
		}
		return result;
	}
	
	/* ------ Other ------ */
	public HttpResult UpdateUserinfoResult (WWW www) {
		HttpResult result = BaseResult(www);
		
		if(HttpResult.ResultType.Sucess == result.resultType) {
			JSONObject resultObj = (JSONObject)result.resultObject;
			EginUser.Instance.UpdateUserWithDict(resultObj.ToDictionary());
		}
		return result;
	}
	
	public HttpResult RoomListResult (WWW www) {
		HttpResult result = BaseResult(www);
		if(HttpResult.ResultType.Sucess == result.resultType) {
			JSONObject resultObject = (JSONObject)result.resultObject;
			List<GameRoom> roomlist = new List<GameRoom>();
			foreach (JSONObject roomInfo in resultObject.list) {
				GameRoom room = new GameRoom(roomInfo.ToDictionary());
				roomlist.Add(room);
			}
			result.resultObject = roomlist;
		}
		return result;
	}

	/* ------ Gift ------ */
	public HttpResult GiftNicknameResult (WWW www) {
		HttpResult result = BaseResult(www);
		if(HttpResult.ResultType.Sucess == result.resultType) {
			result.resultObject = Regex.Unescape(result.resultObject.ToString());
		}
		return result;
	}


	public HttpResult BaseResultSocket (WWW www) {
			HttpResult result = new HttpResult ();
		
			if (www.error != null) {
					EginTools.Log ("Http Failed: " + www.error);
					PlatformGameDefine.playform.swithGameHostUrl ();//切换游戏IP
			} else {
					string tempResultStr = www.text.Trim ();
					EginTools.Log ("base socket:::" + tempResultStr);
		
					JSONObject resultObj = new JSONObject (tempResultStr);
					 
					if (resultObj) {
 						result.resultObject = resultObj;
						result.resultType = HttpResult.ResultType.Sucess;
							 
					} else {
							Debug.Log ("解析出错了---HttpConnect.cs -- 60");
					}
			
 			}
			return result;
	}

	public HttpResult BaseResultSocket2 (WWW www) {
		HttpResult result = new HttpResult ();
		
		if (www.error != null) {
			EginTools.Log ("Http Failed: " + www.error);
			PlatformGameDefine.playform.swithGameHostUrl ();//切换游戏IP
		} else {
			string tempResultStr = www.text.Trim();
			EginTools.Log(tempResultStr);
			
			if ("ok".Equals(tempResultStr)) {
				result.resultType = HttpResult.ResultType.Sucess;
			}else {
				result.resultObject = ZPLocalization.Instance.Get("HttpConnectError");
			}
			
 		}
		return result;
	}
	
	
}

File: Lua/Common/HttpConnect.lua

local this = LuaObject:New()
HttpConnect = this

this.Instance = this

function this:HttpRequestAli( pUrl )
	EginTools.Log(pUrl)
	local www = UnityEngine.WWW.New(pUrl)
	return www 
end

function this:HttpRequest( pUrl,pForm )
	EginTools.Log(pUrl)
	if pForm == nil then
		pForm = UnityEngine.WWWForm.New();
	end
	if pForm ~= nil then
		local tMs = EginTools.nowMinis()
		local tMms = tMs + EginTools.localBeiJingTime
		local tCcode = EginTools.encrypTime(tostring(tMms))
		pForm:AddField('client_code',tCcode)
	end
	local www = nil 
	if pForm == nil then
		www = UnityEngine.WWW.New(pUrl)
	else
		www = UnityEngine.WWW.New(pUrl,pForm)
	end
	return www
end

function this:HttpRequestWithSession( pUrl,pForm )
	EginTools.Log(pUrl)
	local tCookie =nil 
	if EginUser.Instance.session  ~= nil then
		tCookie = EginUser.Instance.session 
	else
		tCookie = ''
	end

	local tRequestHeaders = {}
	tRequestHeaders['Cookie'] = tCookie

	local tHeaders = {}
	tHeaders['Cookie'] = tCookie

	if pForm ==nil then
		pForm = UnityEngine.WWWForm.New()
		pForm:AddField('Cookie',tCookie)
	end
	if pForm ~= nil then
		local tMs = EginTools.nowMinis()
		local tMms = tMs + EginTools.localBeiJingTime
		local tCcode = EginTools.encrypTime(tostring(tMms))
		pForm:AddField('client_code',tCcode)
	end

	local www = Util.GetWWW(pUrl,pForm,tHeaders)   --UnityEngine.WWW.New(pUrl,pForm.data,tHeaders)

	return www
end

-----------Login  ------------

function this:BaseResult( pWWW,pIsSwitchUrl)
	print('******************************BaseResult   Enter  ')
	if pIsSwitchUrl == nil then
		pIsSwitchUrl = true 
	end

	local tResult = HttpResult:New()
	if pWWW.error ~= nil then
		EginTools.Log('Http Failed ====='..pWWW.error)
		if pIsSwitchUrl then
			PlatformGameDefine.playform:swithWebHostUrl(true,Utils.NullObj)
			tResult.isSwitchHost = true
		end
		tResult.resultObject = ZPLocalization.Instance:Get('HttpConnectError')
	else
		local tTempResultStr = Util.SetTrim(pWWW.text)--pWWW.text:Trim()
		EginTools.Log('base ::: '..tTempResultStr)
		local tResultObj =  cjson.decode(tTempResultStr) 
		-- print('>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>')
		-- print(tResultObj)
		-- print(tResultObj['result'])
		-- local tResultObj = JSONObject:New(tTempResultStr) 
		-- if tResultObj.type == JSONObject.Type.NULL then
		-- 	if pIsSwitchUrl then
		-- 		PlatformGameDefine.platform:swithWebHostUrl() 
		-- 		tResult.isSwitchHost = true
		-- 	end
		-- 	tResult.resultObject = ZPLocalization.Instance:Get("HttpConnectError")
		-- else
			local tResultType = nil 
			if tResultObj then 
				if tResultObj['result'] ~= nil then
					tResultType = tResultObj['result']
				end
				if tResultType == 'ok' then
					tResult.resultObject = tResultObj['body']
					tResult.resultType = HttpResult.ResultType.Sucess
				else
					if tResultObj['body'] ~= nil then
						tResult.resultObject = System.Text.RegularExpressions.Regex.Unescape(tResultObj['body'])
					else
						tResult.resultObject = nil
					end
				end
			else
				if pIsSwitchUrl then
					PlatformGameDefine.playform:swithWebHostUrl(true,Utils.NullObj) 
					tResult.isSwitchHost = true
				end
				tResult.resultObject = ZPLocalization.Instance:Get("HttpConnectError")
				print("解析出错了---HttpConnect.lua -- 102")
			end
		-- end
	end
	return tResult
end

--	/* ------ Register ------ */
function this:RegisterResult( pWWW )

	local tResult = HttpResult:New()
	if pWWW.error ~= nil then
		EginTools.Log('Http Failed: '..pWWW.error)
	else
		local tTempResultStr = Util.SetTrim(pWWW.text) --.Trim()
		EginTools.Log(tTempResultStr)
		if tTempResultStr == 'ok' then
			tResult.resultType = HttpResult.ResultType.Sucess
		elseif string.len(tTempResultStr)>0 then
			local tResultObj = cjson.decode(tTempResultStr)
			-- local tResultObj = JSONObject:New(tTempResultStr)
			-- if tResultObj.type == JSONObject.Type.NULL then
			-- print('----------------------------')
			print(tResultObj)
			if tResultObj==nil then
				tResult.resultObject = System.Text.RegularExpressions.Regex.Unescape(tTempResultStr)
			else
				tResult.resultObject = System.Text.RegularExpressions.Regex.Unescape(tResultObj['error'])
			end
		else
			tResult.resultObject = ZPLocalization.Instance:Get('HttpConnectError')
		end
	end
	return tResult
end

function this:UserLoginResult( pWWW )
	local tResult = this:BaseResult(pWWW)
	if HttpResult.ResultType.Sucess == tResult.resultType then
		local tSession = ''
		if pWWW.responseHeaders['SET-COOKIE'] ~= nil  then
			tSession = pWWW.responseHeaders['SET-COOKIE']
		end
		local tResultDict = tResult.resultObject 
		printf(tResultDict)
		-- print('================================fill the dic  ')
		EginUser.Instance:InitUserWithDict(cjson.encode(tResultDict), tSession);
		EginUser.Instance.isGuest = false
	end
	return tResult
end

function this:GuestLoginResult( pWWW )
	local tResult  = this:UserLoginResult(pWWW)
	return tResult
end

function this:UpdateUserinfoResult( pWWW)
	local tResult = this:BaseResult(pWWW)
	-- print('>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> UpdateUserinfoResult')
	-- print(tResult)
	if HttpResult.ResultType.Sucess == tResult.resultType then
		local tResultObj = tResult.resultObject
		EginUser.Instance:UpdateUserWithDict(tResultObj)
	end
	return tResult
end

function this:RoomListResult( pWWW )
	local tResult = this:BaseResult(pWWW)
	if HttpResult.ResultType.Sucess == tResult.resultType then
		local tResultObj = tResult.resultObject
		local tRoomList = {}
		for k,v in tResultObj do
			local tRoom = GameRoom.New(v)
			table.insert(tRoomList ,tRoom)
		end
		tResult.resultObject = tRoomList

	end
	return tResult
end

function this:GiftNicknameResult( pWWW)
	local tResult = this:BaseResult(pWWW)
	if HttpResult.ResultType.Sucess == tResult.resultType then
		tResult.resultObject = System.Text.RegularExpressions.Regex.Unescape(tResult.resultObject)
	end
	return tResult
end

function this:BaseResultSocket(pWWW)
	local tResult = HttpResult:New()
	if pWWW.error ~= nil then
		EginTools.Log('Http Failed ======='..pWWW.error)
		PlatformGameDefine.playform:swithWebHostUrl(true,Utils.NullObj)
	else
		local tTempResultStr = Util.SetTrim(pWWW.text)-- pWWW.text.Trim()
		
		EginTools.Log("base socket::: " ..tTempResultStr)
		-- local tResultObj = JSONObject:New(tTempResultStr)
		 local tResultObj = tTempResultStr
		if tResultObj then
			tResult.resultObject = tResultObj 
			tResult.resultType = HttpResult.ResultType.success
		else
			print("解析出错了---HttpConnect.cs -- 60")
		end
	end
	return tResult
end


function this:BaseResultSocket2(pWWW)
	local tResult = HttpResult:New()
	if pWWW.error ~= nil then
		EginTools.Log('Http Failed : '..pWWW.error)
		PlatformGameDefine.playform:swithWebHostUrl(true,Utils.NullObj)
	else
		local tTempResultStr =Util.SetTrim(pWWW.text)
		EginTools.Log(tTempResultStr)
		if tTempResultStr == 'ok' then	
			tResult.resultType = HttpResult.ResultType.Sucess
		else		
			tResult.resultObject = ZPLocalization.Instance:Get('HttpConnectError')
		end
	end
	return tResult
end
